perf: Improve get/set lookup times

obastemur opened the pull request:
- Do not force create / cache PropertyString at the first place. PropertyRecord is what we need most of the time.
- Extend LiteralStringWithPropertyStringPtr with PropertyRecord (40 -> 48 bytes) and align other 40 bytes string object similarly. Long story short.. No actual extra memory usage.

akroshg commented:
@MikeHolman can you take a look at this one?

MikeHolman commented:
Can you validate perf using //onecirro? I added this for some frameworks in Speedometer. I'm not 100% sure if this particular part was necessary or not, but I suspect there will be perf impact.

MikeHolman commented:
If you are coming here often, I think THAT is what you should focus on. This is a problem I've also been wrestling with a lot in the last year.

Where are the strings coming from for this? Is there any way we can know ahead of time that they should be PropertyString, etc.? IMO optimizing this code path is suppressing a symptom, not actually solving the problem.

MikeHolman commented:
One thing I was wondering is if there was a way to add a cache to the PropertyRecord *itself* we might be able to solve this problem a bit better. More generally, it might be worth setting up a meeting with you, me, @LouisLaf, and @curtisman to spend some time brainstorming, because this is a problem that has been biting us a lot recently (in both AcmeAir and Speedometer).

obastemur commented:
>  I added this for some frameworks in Speedometer.
> If you are coming here often, I think THAT is what you should focus on

@MikeHolman `this` / `that` are not clear to me. At least not sure what they are pointing out. Was there a Github editor issue? Could you please re-comment?

I was going to send a similar PR probably a week later but @akroshg pinged me on a perf regression with `Edge_fastdom_expando_setget` and RI.. So, I combine couple of things I was planning to achieve on this PR.

I already did run the benchmarks (not Speedometer yet). Kraken `audio-oscillator` ~5% better and tested a simplified version of `Edge_fastdom_expando_setget` and saw 40% improvement. No regression on the other benchmarks. Not sure why it should be? **Also saw a decent transient memory improvement on acme-air. This was actually my starting point.**

Recently, I had implemented `GetPropertyRecord` on JavascriptString et al. and `JsCreateString` on ChakraCore was creating `LiteralStringWithPropertyStringPtr`.. So, while we were benefiting the cached PropertyString, I realized we actually need `PropertyRecord` but not `PropertyString` most of the time. However, there are cases we also need PropertyString and luckily we have enough space under the `LiteralStringWithPropertyStringPtr` to accomodate both for now.

There is more to do..

I'm trying to understand why we need PropertyString at this point.. OR both PropertyRecord and PropertyString. (add LiteralStringWithPropertyStringPtr on top of everything.. ) If this is what you were referring, IMO yes! we may not need all of these?

However, I'm trying to collect more data and understand how all these string variations work together or collapse.

This PR only -> Better memory + Better perf ? If you see a potential regression on any of these, could you please point it out on the changes?

MikeHolman commented:
I mean these paths in GetElementI/SetElementI are the slow paths, if you are here it means you missed all of our inline caches and/or didn't have a PropertyString.

And if you are spending time on the non-PropertyString path, why are you spending the time here? I think solving the problem of why you are in Get/SetElementI to begin with is the more important problem.

Where I think there is possible regression is with the GetOrAddPropertyString change to GetPropertyString. Basically the "Add" is so that later uses of the string will be optimized, even at a cost here. And the point of the PropertyString is that it has an InlineCache, so property accesses using the string can be faster.

I think there are a lot of places where we would want (but don't/can't get) a PropertyString, and so maybe we should consider whether there is a better approach.

obastemur commented:
> I mean these paths in GetElementI/SetElementI are the slow paths, if you are here it means you missed all of our inline caches and/or didn't have a PropertyString.

This is only a small part of the change and it is the one related to regression.

> Where I think there is possible regression is with the GetOrAddPropertyString change to GetPropertyString. Basically the "Add" is so that later uses of the string will be optimized, even at a cost here. 

I don't have a data supporting this. Can you point out the `GetPropertyString` call that was benefiting that already? and not it can't?

MikeHolman commented:
>                     propertyString = propStr->GetOrAddPropertyString();

might be worth having the same logic here

---
Refers to: lib/Runtime/Language/JavascriptOperators.cpp:3750 in e98b622. [](commit_id = e98b62255b79691f6bcae7dd6d336708b23509d7, deletion_comment = False)

obastemur commented:
Chatted with @akroshg @MikeHolman offline.
`SetElementIHelper` is a common place to setup the `PropertyString` during the warm-up. Oppositely, @akroshg 's regression test was based on unique refs.. Updated PR to meet two scenarios on the middle ground. Create PropertyString only if we already had created a propertyRecord cache. So, we will still keep creation of one-use-only propertyStrings.

obastemur commented:
> might be worth having the same logic here

I was going to but that place strictly depend on PropertyString unlike this one. It can be a thing to look at but skipped for now
